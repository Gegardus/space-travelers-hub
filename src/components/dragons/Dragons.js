import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';
import {
  getDragonsFromAPI,
  reserveDragon,
  cancelDragonReservation,
} from '../../redux/dragons/dragons';

const Dragons = () => {
  const dispatch = useDispatch();
  const dragonsData = useSelector((state) => state.dragons);
  const getDragons = () => {
    if (dragonsData.length === 0) {
      dispatch(getDragonsFromAPI());
    }
  };
  useEffect(() => {
    getDragons();
  }, []);

  return (
    <div className="dragons-list">
      {dragonsData.map((dragon) => (
        <div key={dragon.dragon_id} className="dragon">
          <div className="dragon-img">
            <img src={dragon.dragon_img} alt={dragon.dragon_name} />
          </div>
          <div className="dragon-description">
            <h3>{dragon.dragon_name}</h3>
            <p>
              {dragon.reserved && (
                <Badge bg="info" className="reserved-badge">
                  Reserved
                </Badge>
              )}
              {dragon.dragon_description}
            </p>
            {dragon.reserved ? (
              <Button
                variant="outline-secondary"
                onClick={() => dispatch(cancelDragonReservation(dragon.dragon_id))}
              >
                Cancel Reservation
              </Button>
            ) : (
              <Button
                variant="primary"
                onClick={() => dispatch(reserveDragon(dragon.dragon_id))}
              >
                Reserve Dragon
              </Button>
            )}
          </div>
        </div>
      ))}
    </div>
  );
};

export default Dragons;
