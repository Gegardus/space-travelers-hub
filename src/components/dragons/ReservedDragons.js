import React from 'react';
import { useSelector } from 'react-redux';

const ReservedDragons = () => {
  const selectedDragons = useSelector((state) => state.dragons);
  const reservedDragons = selectedDragons.filter((dragon) => dragon.reserved === true);
  return (
    <ul className="profile-container">
      <h4>My Dragons</h4>
      {
 reservedDragons.map((dragon) => (
   <li key={dragon.dragon_id}>
     <p>{dragon.dragon_name}</p>
   </li>
 ))
}
    </ul>
  );
};

export default ReservedDragons;
