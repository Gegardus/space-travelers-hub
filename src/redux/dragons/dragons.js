const ADD_DRAGON = 'spacetravelers-hub/redux/dragons/ADD_DRAGON';
const dragonURL = 'https://api.spacexdata.com/v3/dragons';

const initialState = [];

export const addDragon = (dragons) => ({
  type: ADD_DRAGON,
  dragons,
});

export const getDragonsFromAPI = () => (async (dispatch) => {
  const response = await fetch(dragonURL);
  const data = await response.json();
  const dragonsArr = [];
  data.forEach((d) => {
    const dragon = {
      dragon_id: d.id,
      dragon_name: d.dragon_name,
      dragon_description: d.description,
      dragon_img: d.flickr_images[0],
    };
    dragonsArr.push(dragon);
  });
  dispatch(addDragon(dragonsArr));
});

const dragonReducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_DRAGON:
      return [...state, ...action.dragons];
    default:
      return state;
  }
};
export default dragonReducer;
