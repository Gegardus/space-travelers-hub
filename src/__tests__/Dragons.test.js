import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import store from '../redux/configureStore';
import Dragons from '../components/dragons/Dragons';
import { addDragon, cancelDragonReservation, reserveDragon } from '../redux/dragons/dragons';

describe('Dragons Component test', () => {
  it('Renders Rocket Page', () => {
    const app = renderer.create(
      <Provider store={store}>
        <Dragons />
      </Provider>,
    ).toJSON();
    expect(app).toMatchSnapshot();
  });

  it('has a text in the DOM', () => {
    const { container } = render(<Provider store={store}><Dragons /></Provider>);
    expect(container.getElementsByClassName('dragons-list').length).toBe(1);
  });
  it('Returns an addDragon() action object ', () => {
    expect(addDragon([])).toStrictEqual({ dragons: [], type: 'spacetravelers-hub/redux/dragons/ADD_DRAGON' });
  });
  it('Returns a reserveDragon() action object ', () => {
    expect(reserveDragon('5107')).toStrictEqual({ dragonId: '5107', type: 'spacetravelers-hub/redux/dragons/RESERVE_DRAGON' });
  });
  it('Returns a cancelDragonReservation() action object ', () => {
    expect(cancelDragonReservation('5107')).toStrictEqual({ dragonId: '5107', type: 'spacetravelers-hub/redux/dragons/CANCEL_RESERVATION' });
  });
});
