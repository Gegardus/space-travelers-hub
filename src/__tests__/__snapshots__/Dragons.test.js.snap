// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dragons Component test Renders Rocket Page 1`] = `
<div
  className="dragons-list"
/>
`;
